serf/event.go

```diff
@@ -156,13 +158,53 @@ func (q *Query) Respond(buf []byte) error {
 		return fmt.Errorf("response exceeds limit of %d bytes", q.serf.config.QueryResponseSizeLimit)
 	}
 
-	// Send the response
-	addr := net.UDPAddr{IP: q.addr, Port: int(q.port)}
-	if err := q.serf.memberlist.SendTo(&addr, raw); err != nil {
-		return err
+	// Relay the response through up to QueryResponseRelayLimit nodes
+	members := q.serf.Members()
+	relayMembers := kRandomMembers(q.serf.config.QueryResponseRelayLimit, members, func(m Member) bool {
+		return m.Status != StatusAlive && m.ProtocolMax >= 5
+	})
+
+	for _, m := range relayMembers {
+		relayAddr := net.UDPAddr{IP: m.Addr, Port: int(m.Port)}
+		if err := q.serf.memberlist.SendTo(&relayAddr, raw); err != nil {
+			return err
+		}
 	}
 
-	// Clera the deadline, response sent
+	// Clear the deadline, response sent
 	q.deadline = time.Time{}
 	return nil
 }
+
+// kRandomMembers selects up to k members from a given list, optionally
+// filtering by the given filterFunc
+func kRandomMembers(k int, members []Member, filterFunc func(Member) bool) []Member {
```

Comment:
I think this is fine as-is but it's interesting if you upgrade a few nodes of a large cluster I bet this'll come up empty handed sometimes. I think that's actually a good feature because you won't overwhelm a small amount of nodes with relays as you are upgrading things, and it'll get progressively better at finding nodes once there are a lot of them that support relay.